genesis/engine/solvers/rigid/gjk_decomp.py

```diff
@@ -0,0 +1,3394 @@
+                if i == 1:
+                    j_s1, j_s2, j_s3 = i_s1, i_s3, i_s4
+                    k_1, k_2, k_3 = 0, 2, 3
+                elif i == 2:
+                    j_s1, j_s2, j_s3 = i_s1, i_s2, i_s4
+                    k_1, k_2, k_3 = 0, 1, 3
+                elif i == 3:
+                    j_s1, j_s2, j_s3 = i_s1, i_s2, i_s3
+                    k_1, k_2, k_3 = 0, 1, 2
+
+                if not scs[i]:
+                    _lambda2d = self.func_gjk_subdistance_2d(i_b, j_s1, j_s2, j_s3)
+                    closest_point = self.func_simplex_vertex_linear_comb(i_b, 2, j_s1, j_s2, j_s3, 0, _lambda2d, 3)
+                    d = closest_point.dot(closest_point)
+                    if d < dmin:
+                        dmin = d
+                        _lambda.fill(0.0)
+                        _lambda[k_1] = _lambda2d[0]
+                        _lambda[k_2] = _lambda2d[1]
+                        _lambda[k_3] = _lambda2d[2]
+
+        return _lambda
+
+    @ti.func
+    def func_gjk_subdistance_2d(self, i_b, i_s1, i_s2, i_s3):
+
+        _lambda = ti.math.vec4(0, 0, 0, 0)
+
+        # Project origin onto affine hull of the simplex (triangle)
+        proj_o, proj_flag = self.func_project_origin_to_plane(
+            self.gjk_simplex_vertex[i_b, i_s1].mink,
+            self.gjk_simplex_vertex[i_b, i_s2].mink,
+            self.gjk_simplex_vertex[i_b, i_s3].mink,
+        )
+        if proj_flag:
+            # If projection failed because the zero normal,
+            # project on to the first edge of the triangle
+            _lambda = self.func_gjk_subdistance_1d(i_b, i_s1, i_s2)
+        else:
+            # We should find the barycentric coordinates of the projected point,
+            # but the linear system is not square:
+            # [ s1.x, s2.x, s3.x ] [ l1 ] = [ proj_o.x ]
+            # [ s1.y, s2.y, s3.y ] [ l2 ] = [ proj_o.y ]
+            # [ s1.z, s2.z, s3.z ] [ l3 ] = [ proj_o.z ]
+            # [ 1,    1,    1,   ] [ ?  ] = [ 1.0 ]
+            # So we remove one row before solving the system
+            # We exclude the axis with the largest projection of the simplex
+            # using the minors of the above linear system.
+            s1 = self.gjk_simplex_vertex[i_b, i_s1].mink
+            s2 = self.gjk_simplex_vertex[i_b, i_s2].mink
+            s3 = self.gjk_simplex_vertex[i_b, i_s3].mink
+
+            m1 = s2[1] * s3[2] - s2[2] * s3[1] - s1[1] * s3[2] + s1[2] * s3[1] + s1[1] * s2[2] - s1[2] * s2[1]
+            m2 = s2[0] * s3[2] - s2[2] * s3[0] - s1[0] * s3[2] + s1[2] * s3[0] + s1[0] * s2[2] - s1[2] * s2[0]
+            m3 = s2[0] * s3[1] - s2[1] * s3[0] - s1[0] * s3[1] + s1[1] * s3[0] + s1[0] * s2[1] - s1[1] * s2[0]
+
+            m_max = 0.0
+            absm1, absm2, absm3 = ti.abs(m1), ti.abs(m2), ti.abs(m3)
+            s1_2d, s2_2d, s3_2d = gs.ti_vec2(0, 0), gs.ti_vec2(0, 0), gs.ti_vec2(0, 0)
+            proj_o_2d = gs.ti_vec2(0, 0)
+
+            if absm1 >= absm2 and absm1 >= absm3:
```

Comment:
You should be able to factorize the contain of these 3 conditions, with indices to extract that depends on your conditions.